### PR TITLE
HTTP/2 Graceful Shutdown Timeout

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
If any streams are still active the graceful shutdown code will wait until they are all closed before the connection is closed. In some situations this event may never occur, and thus a timeout should be supported so the socket can be closed even if all streams haven't been closed.

Modifications:
- Add a configurable timeout for when the graceful shutdown process is attempted.
- Update unit tests to be faster, and use this graceful timeout

Result:
Local endpoint can protect from local or remote issues which prevent the channel from being closed during the graceful shutdown process.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler - PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 1]
```diff
@@ -53,11 +57,13 @@
  */
 public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http2LifecycleManager {
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(Http2ConnectionHandler.class);
+
     private final Http2ConnectionDecoder decoder;
     private final Http2ConnectionEncoder encoder;
     private final Http2Settings initialSettings;
     private ChannelFutureListener closeListener;
     private BaseDecoder byteDecoder;
+    private long gracefulShutdownTimeoutMillis = 5000;
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I think the default for this should be something much bigger to support large streams ... or just default to MAX_INT.  WDYT?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 2]
```diff
@@ -716,14 +745,30 @@ private static void processGoAwayWriteResult(final ChannelHandlerContext ctx, fi
...
+            timeoutTask = null;
+        }
+
+        ClosingChannelFutureListener(ChannelHandlerContext ctx, ChannelPromise promise, long timeout, TimeUnit unit) {
+            this.ctx = ctx;
+            this.promise = promise;
+            timeoutTask = ctx.executor().schedule(new OneTimeTask() {
+                @Override
+                public void run() {
+                    ClosingChannelFutureListener.this.ctx.close(ClosingChannelFutureListener.this.promise);
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: could you just use the parameter `ctx` here (mark it as final)?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java [hunk 3]
```diff
@@ -265,7 +268,7 @@ public void clientRequestOneDataFrame() throws Exception {
                 @Override
                 public void run() {
                     frameWriter.writeHeaders(ctxClient(), 3, http2Headers, 0, false, newPromiseClient());
-                    frameWriter.writeData(ctxClient(), 3, content.retain(), 0, true, newPromiseClient());
+                    frameWriter.writeData(ctxClient(), 3, content.duplicate().retain(), 0, true, newPromiseClient());
```
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java]: why did you change all of these to `duplicate`?

**Comment by nmittler** [comment]
@Scottmitch Just finished a first pass .. looks good!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Done

File: codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java]: Otherwise equality comparisons fail.  With the new `Local*Channel` the ByteBuf indexes are updated before the mockito matchers are applied and our output is not captured because we have a eq(..) comparison for the bytebuf argument on the matchers.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: I can increase, but what about something more practical...say 15 minutes?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 4]
```diff
@@ -53,11 +57,14 @@
  */
 public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http2LifecycleManager {
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(Http2ConnectionHandler.class);
+    private static final int DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS = 15 * 60 * 1000; // 15 minutes
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @nmittler - Does 15 minutes work for you?

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: 15 minutes seems reasonable.  @ejona86 any thoughts?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java [hunk 5]
```diff
@@ -47,6 +63,13 @@ public Http2Error error() {
     }
 
     /**
+     * If a connection error should attempt a graceful shutdown.
+     * @return {@code true} if a connection error should attempt a graceful shutdown. {@code false} otherwise.
+     */
+    public boolean gracefulShutdown() {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java]: @nmittler - Please note this change (and how it is used). A unit test was failing when the stream ids were exhausted...which lead me to look at this code path...and I realized in this case we should be attempting a graceful shutdown.  I could have done an `instanceof` check but figured it was more general to put support for this in `Http2Exception` (not this is not supported and always false for `StreamError`s.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 4 above)
Comment (ejona86) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: 15 minutes seems _fine_. Very long if you aren't aware of it though (which is my experience with the graceful shutdown: "not aware of it and getting bitten").

I would have assumed a figure more like 5-30 seconds to just let normal HTTP requests to complete. I wonder if it would be better to default to 0 so applications know what they are getting into and that essentially any number chosen here will be "wrong".

But since any number chosen here will be "wrong," I'm not really concerned. 15 minutes works.

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @ejona86 - I would like to be more aggressive as well (previously set to 5 seconds).

@nmittler - Would 1 minute or 30 seconds work for you?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java (hunk 5 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java]: nit: Make `shouldShutdownGracefully`?  Also I wonder if it would make sense to make this an enum:

```java
enum ShutdownMode {
    NO_SHUTDOWN_REQUIRED,
    SHUTDOWN_GRACEFULLY,
    SHUTDOWN_NOW
}
```

The default for stream errors would be `NO_SHUTDOWN_REQUIRED` and the default for connection errors would be `SHUTDOWN_GRACEFULLY`.  WDYT?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 4 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: @Scottmitch Yeah now that I think about it, I think this makes sense.  +1 to 30 seconds.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapterTest.java]: ah ok, thanks for the explanation.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java (hunk 5 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java]: I'm a bit hesitant to introduce an enum because it conflicts a bit with the `isStreamError` logic. If we introduce this enum its almost like we should transition over to using this enum instead of checking the type of exception, but then there are already some semantics associated with the protocol and other assumptions we make based upon types.

Basically what I want to capture in the method name is something like the following: `ifAShutdownIsWarrantedThenDoAGracefulShutdown`. So it is clear the logic to execute a shutdown is not necessarily made at time time of exception creation. Maybe `shouldShutdownBeGraceful`?

**Comment by Scottmitch** [comment]
@nmittler - I have resolved all comments, but we may still want to discuss https://github.com/netty/netty/pull/4106#discussion_r37551548.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java (hunk 5 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Exception.java]: Actually I think that having both this and `isStreamError` is a bit confusing.  IMO this discussion is orthogonal to our exception hierarchy .. I don't think we need to change any of that. I actually think that having an enum like this (maybe `ShutdownHint` would be a better name?) could make our shutdown decision making more straight forward. How hard do you think it would be to change the `isStreamError` handling logic over to this?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 6]
```diff
@@ -53,11 +57,14 @@
  */
 public class Http2ConnectionHandler extends ByteToMessageDecoder implements Http2LifecycleManager {
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(Http2ConnectionHandler.class);
+    private static final int DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS = 30 * 1000; // 30 seconds
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: nit: just use TimeUnit?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: Done

**Comment by Scottmitch** [comment]
@nmittler - I implemented you suggestion. I currently left some of the "graceful" terminology because it is extensible, and can be overriden to change behavior.  Let me know if you can think of a better name.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java [hunk 7]
```diff
@@ -551,7 +586,17 @@ protected void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Htt
         if (http2Ex == null) {
             http2Ex = new Http2Exception(INTERNAL_ERROR, cause.getMessage(), cause);
         }
-        goAway(ctx, http2Ex).addListener(new ClosingChannelFutureListener(ctx, ctx.newPromise()));
+
+        ChannelPromise promise = ctx.newPromise();
+        ChannelFuture future = goAway(ctx, http2Ex);
+        switch (http2Ex.shutdownHint()) {
+        case SOFT_SHUTDOWN:
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: nit: Since "graceful` doesn't really refer to the HTTP/2 definition here, maybe we should also call this "graceful" for now ... since it basically means the same thing as the other uses of "graceful".  WDYT?

**Comment by nmittler** [comment]
@Scottmitch just one minor comment ... otherwise LGTM

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java (hunk 7 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java]: SGTM. Done.

**Comment by Scottmitch** [comment]
@nmittler - Updated enum and comment describing the new value.

**Comment by nmittler** [comment]
@Scottmitch ship it! :)

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (08477ea) master (6a4a5f5)
